tracer/src/Datadog.Trace/Debugger/Instrumentation/Registry/MethodMetadataRegistry.cs

```diff
@@ -0,0 +1,136 @@
+                    ref _instance,
+                    ref _instanceInitialized,
+                    ref _instanceLock);
+            }
+        }
+
+        /// <summary>
+        /// Tries to create a new <see cref="MethodMetadataInfo"/> at <paramref name="index"/>.
+        /// </summary>
+        /// <param name="targetObject">The target object</param>
+        /// <param name="index">The index of the method inside <see cref="EverGrowingRegistry{TPayload}.Items"/></param>
+        /// <param name="methodHandle">The handle of the executing method</param>
+        /// <param name="typeHandle">The handle of the type</param>
+        /// <param name="asyncKickOffInfo">The info for the async kickoff method</param>
+        /// <returns>true if succeeded (either existed before or just created), false if fails to create</returns>
+        public bool TryCreateAsyncMethodMetadataIfNotExists<TTarget>(TTarget targetObject, int index, in RuntimeMethodHandle methodHandle, in RuntimeTypeHandle typeHandle, AsyncHelper.AsyncKickoffMethodInfo asyncKickOffInfo)
+        {
+            if (IsIndexExists(index))
+            {
+                return true;
+            }
+
+            // Create a new one at the given index
+            lock (ItemsLocker)
+            {
+                if (IsIndexExists(index))
+                {
+                    return true;
+                }
+
+                EnlargeCapacity(index);
+
+                var method = MethodBase.GetMethodFromHandle(methodHandle, typeHandle);
+
+                if (Log.IsEnabled(Vendors.Serilog.Events.LogEventLevel.Debug))
+                {
+                    Log.Debug($"{nameof(MethodMetadataRegistry)}.{nameof(TryCreateAsyncMethodMetadataIfNotExists)}: Creating a new metadata info for Async method = {method}, index = {index}, Items.Length = {Items.Length}");
+                }
+
+                if (method == null)
+                {
+                    return false;
+                }
+
+                var targetType = targetObject == null ? Type.GetTypeFromHandle(typeHandle) : targetObject.GetType();
+                Items[index] = MethodMetadataInfoFactory.Create(method, targetType, asyncKickOffInfo);
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Tries to create a new <see cref="MethodMetadataInfo"/> at <paramref name="index"/>.
+        /// </summary>
+        /// <param name="index">The index of the method inside <see cref="EverGrowingRegistry{TPayload}.Items"/></param>
+        /// <param name="methodHandle">The handle of the executing method</param>
+        /// <param name="typeHandle">The handle of the type</param>
+        /// <returns>true if succeeded (either existed before or just created), false if fails to create</returns>
+        public bool TryCreateNonAsyncMethodMetadataIfNotExists(int index, in RuntimeMethodHandle methodHandle, in RuntimeTypeHandle typeHandle)
```

Comment:
This method is almost identical to `TryCreateAsyncMethodMetadataIfNotExists`, is there a way to reduce code duplication here?

Reply:
I was thinking about it initially, these two functions play part in a really hot path (entrance of every line & method probes) so I decided to keep like that for the being since I'm not sure refactoring this piece to save 4-5 lines of code really worth it (taking into account that we talk about a hot path)